### ISSUE TITLE
Add pyproject.toml support

### ISSUE THREAD

**Issue by seanlaw** [TDAmeritrade/stumpy#343]
See [this blogpost](http://ivory.idyll.org/blog/2021-transition-to-pyproject.toml-example.html)

**Comment by davidrhallartificialeducation** [comment]
I will get to work on this issue

**Comment by seanlaw** [comment]
@davidrhallartificialeducation Thank you for your interest and welcome to the STUMPY community. Before you get too far, I think that it might be a good idea for us to discuss the overall goal and any concerns. Personally, I have no experience with `pyproject.toml` but, from what I've been hearing/seeing, people seem to be moving in that direction. My biggest concern is maintaining backwards compatibility with other Python versions (3.6+). It would be great if we could start with a checklist of things that need to be done. I'd love to hear your thoughts

**Comment by davidrhallartificialeducation** [comment]
Absolutely, that'd be great! I'm a new contributor in general to open source project so I definitely appreciate any mentorship on the process.  Currently I'm getting myself familiar with the project, black, and flake8.  From what I've been reading at the blog post it looks like pyproject.toml allow's us to version the build tools required to run STUMPY.

**Comment by seanlaw** [comment]
> Absolutely, that'd be great! I'm a new contributor in general to open source project so I definitely appreciate any mentorship on the process. Currently I'm getting myself familiar with the project, black, and flake8. From what I've been reading at the blog post it looks like pyproject.toml allow's us to version the build tools required to run STUMPY.

I just came across [this write-up](https://laptrinhx.com/brett-cannon-what-the-heck-is-pyproject-toml-3531099003/) by Brett Cannon (a Python steering committee member) that may be useful.

To be clear, this Github issue should really be asking "should we add pyproject.toml to STUMPY?" Perhaps, after an informed discussion, the answer may be "No" or "Not yet". So, just throwing it out there that this is more about information gathering and we may find that the STUMPY codebase is simple enough and doesn't actually benefit from `pyproject.toml`. Maybe you'd be willing to help us assess the following:

1. What kinds of projects typically benefit from having a `pyproject.toml`
2. Which of those benefits apply to STUMPY?

Again, don't feel pressured and please feel free to "pass" if your intent is to contribute code rather than to gather information. I'm sure that we can find other opportunities for you too!

**Comment by davidrhallartificialeducation** [comment]
@seanlaw 

# Should We Add Pyproject.toml support

### Arguments for adding it
* configuration files are better than scripts
* a few standard configuration files are better than many
* declarative/static is better than procedural/dynamic

### Benefits
* Can rely on specific versions of setuptools and ‘wheel’
* Can substitute build tools for others (i.e. setuptools for poetry)
* No assumptions needed for what's required to build the project
* Build Isolation as tools are installed in a virtual environment

### What's Required

1. Create a file called pyproject.toml and add the following where X is the setuptools required
```python
[build-system]
requires = ["setuptools >= X.X.X", "wheel"]
build-backend = "setuptools.build_meta"
```

2. Transfer all setup.py configurations to setup.cfg

3. replace setup.py code with:
```python
#!/usr/bin/env python

import setuptools

if __name__ == "__main__":
    setuptools.setup()
```

**Comment by seanlaw** [comment]
Thanks, @davidrhallartificialeducation! So, right now, STUMPY doesn't depend on any particular version of `setuptools` or `wheel`. Since STUMPY is essentially "pure Python" and not specific to any particular architecture, we don't really "build" anything or have specific processes for different architectures. Based on the assessment that you've provided, we'd actually end up with more files rather than less files. I was hoping that `pyproject.toml` might help us reduce complexity but maybe we're one of the rare cases where that may not be the case. I'm starting to lean towards "it's probably not worth it right now and maybe we could revisit this when there is wider adoption of pyproject.toml". What do you think?

**Comment by davidrhallartificialeducation** [comment]
Yes, I agree.  It's been awhile since I've done python and can't remember a lot about the `setuptools` or `wheel`.  The question I would ask is when someone runs `pip install stumpy` does it then need to be built on their local machine? If it does, then I believe this allows them to choose tools to use to "build" it.  Otherwise if not, I don't see much use at the moment.

**Comment by seanlaw** [comment]
> The question I would ask is when someone runs pip install stumpy does it then need to be built on their local machine? If it does, then I believe this allows them to choose which tools to use to "build" it. Otherwise if not, I don't see much use at the moment.

The answer is "no". Thanks to `numba` (a JIT compiler), nothing is built/compiled until run-time and, in STUMPY's case, `pip install stumpy` is just there to copy over the `stumpy` directory to the right place. I think we lucked out when we chose `numba` over `cython`, which greatly simplified or eliminated an explicit "build process". 

@davidrhallartificialeducation This was very helpful to talk through!
